### ISSUE TITLE
Support abstraction over asynchronous computations, revisited

### ISSUE THREAD

**Issue by crusso** [dfinity/motoko#2945]
This a sketch how we might support abstraction of asynchronous code without compromising the "always" commit semantics of `await`, unlike previous proposals (https://github.com/dfinity/motoko/issues/1482).

The basic idea is to introduce a synchronous `lazy T` type  constructor that subtypes `async T` by providing an additional, synchronous `force` operation for forcing the computation delayed by `lazy e`.

Types:

Formation:
```
T ok
---------------
lazy T ok
```

(no restriction to `T` shared).

# Subtyping: 

Covariance
```
  T  <: U
  ----------------------
  lazy T <: lazy U
```

Async inclusion (optional)

```
  T shared
  -------------------------
  lazy T <: async T 
```

# Terms:

Introduction

```
C, async |- e : T
-----------------
C |- lazy e : lazy t
```

Elimination

```
C, async |- e : lazy t
---------------
C, async |- force e : t
```

Admissable rules (by virtue of conditional subtyping):

```
e : lazy T    T shared
-------------------
C, async |- await e : T
```

(Silly) example:

```
private func flip() : lazy Bool  = lazy { 
  let blob = await Random.rand();
  blob[0] & 1 == 1 
};

shared func oldCoin() : async Bool {
    await flip();  // does two real awaits.
}

shared func newCoin() : async Bool {
    force flip();  // only does one real await (that done by flip()).
}
```

This sketch ignores the issues of scoped awaits, but the idea would be that lazy<T> is indexed by an implicit scope parameter and is only well-formed (an introducible) at the current index in an async context (much as we deal with async types now).

Remark: Why is `lazy T ` a subtype of `async T` and not the other way around? Because `lazy T` supports `await` but also offers a additional observation (`force`).

# Questions: 

1. If we support subtyping, it seems sensible for `lazy T` to be stateful with `force` executing the computation exactly once. Another design might be to drop the subtyping, rename `lazy` to `delay` (or even `sync`) and repeat the computation each time the value is forced. This would make objects of that type values, that can be composed without incurring effects. But can we think of any motivating examples?

2. It would be nice to make the subtyping unconditional, for any `T`, by relaxing the restriction (`T` shared) on async `T` and imposing in on the codomains of shared functions instead. That would allow more useful composition and abstraction over async types and their contents and removes an artificial restriction that is probably not required by the implementation. This relaxation has been mooted before.

# Sugar:

```
private func flip() : lazy Bool  { 
  let blob = await Random.rand();
  blob[0] & 1 == 1 
};
```

**Comment by rossberg** [comment]
Hm, can you illustrate what introducing a second variant of async type achieves? Wouldn't a simpler version of this proposal consist of just adding a `possibly_await` operator on async itself? What does splitting that off into a new subtype accomplish, and what drawbacks does it avoid?

**Comment by nomeata** [comment]
Isn't it already too late by the time the code reaches `possibly_await` or `await`? The spawning of a separate (self-)call that we want to avoid in code abstracting over `await` already happens during the evaluation of `await …`, so a new intro form is also needed, right? (Maybe it's ok to have the extra intro and elim forms, but still only one type?)

**Comment by crusso** [comment]
Did you mean:
> Isn't it already too late by the time the code reaches `possibly_await` or `await`? The spawning of a separate (self-)call that we want to avoid in code abstracting over `await` already happens during the evaluation of ~`await …`~ `async ...`, so a new intro form is also needed, right? (Maybe it's ok to have the extra intro and elim forms, but still only one type?)

**Comment by nomeata** [comment]
Yes, thanks. Edited.

**Comment by crusso** [comment]
Just for comparison, https://github.com/dfinity/motoko/issues/1482 propose adding an eager `do async e` block, that would synchronously executes until the first await that suspends and could be used instead of `async e` to avoid the overhead of scheduling. 

So the proposal here is slightly different. Whether its better, I'm not sure. Maybe its nice to know that the effects will only happen at force and await, rather that depending on how the async is introduced.

**Comment by rossberg** [comment]
Yes, we already discussed the option of having a may-commit operator of some form. Doesn't the current proposal still have the same problem, then? My view was that commits and await syntax ought to maintain a 1:1 relation, reflecting that async is a monad and await its bind operator. This proposal avoids the issue of not performing a commit where `await` is used, but it still allows a commit to happen in places (of a function) where `await` is _not_ used. Effectively, the `force` operator circumvents the monad's bind operator. So from my perspective, it still obscures atomicity rules and worsens the concurrency footgun.

Weren't there plans to address the performance issue on the platform level by executing self calls directly?

**Comment by crusso** [comment]
First of all, isn't it a bit of stretch claiming that async is a monad? A monad, in my book, describes an effectful computation as an inert value that can be composed sans side-effect until eliminated. Async is manifestly not that.

The commit points are only important when they affect the state of other concurrent operations. I think we should provide users with the ability to write code that can still be performant and may or may not commit, when they know what they are doing and the commit makes no difference.

The optimization of scheduling self calls directly is bound to run into limits and, in my mind, compromises the fairness of the scheduler. I'm still reluctant to rely on that.

**Comment by rossberg** [comment]
Yes, it's more second-class in Motoko, but I'd think it still forms a monad in the semantics. At least that's a common model for async, and one that provides nice reasoning principles.

> The commit points are only important when they affect the state of other concurrent operations.

If a rollback is unintentionally partial, then you might be left with incoherent state forever, independent of whether this is observed by concurrent or sequential operations.

> The optimization of scheduling self calls directly is bound to run into limits and, in my mind, compromises the fairness of the scheduler. I'm still reluctant to rely on that.

Perhaps. But between this possibility and the general advice to cleanly separate code into async (communication) and sync (internal) layers (like with the use of the IO monad in Haskell), do we have evidence of this being enough of a remaining problem in practice? After all, this addition would be a significant complication to the language that is likely to be more error-prone, increase dev confusion and reduce code clarity. So, I'd want to be certain that we are not just putting it in as a sort of premature optimisation to the design.

**Comment by nomeata** [comment]
The separation in Haskell would not work at all if you couldn't abstract over patterns that involve IO (or is that your point? I'm a bit flight lagged)

**Comment by christoph-dfinity** [comment]
Implemented as `async*/await*` in https://github.com/dfinity/motoko/pull/3609
